### ISSUE TITLE
Hot reloading does not work with @babel/plugin-transform-modules-commonjs

### ISSUE THREAD

**Issue by michaelgmiller1** [pmmmwh/react-refresh-webpack-plugin#120]
When I use the `@babel/plugin-transform-modules-commonjs` plugin and change the repo to use an object export for `App`, hot reloading breaks with the following error:

```
[HMR] Cannot apply update. Need to do a full reload!
VM1440 react_devtools_backend.js:6 [HMR] Error: Aborted because ./src/App.jsx is not accepted
Update propagation: ./src/App.jsx -> ./src/index.js -> 1
    at hotApplyInternal (http://localhost:8080/main.js:537:30)
    at hotApplyInternal (http://localhost:8080/main.js:770:20)
    at hotApply (http://localhost:8080/main.js:391:19)
    at http://localhost:8080/main.js:366:22
```

I've developed a minimum repro here: https://github.com/michaelgmiller1/react-refresh-webpack-plugin/tree/transform-repro

After patching, you can run the `webpack-dev-server` example and modify the `App.jsx` file to reproduce.

Let me know if you need any more info!

Thanks!
Mike

**Comment by michaelgmiller1** [comment]
anything I can do to help investigate this? super excited to adopt react refresh in our app!

**Comment by brandonpapworth** [comment]
👍

**Comment by michaelgmiller1** [comment]
@pmmmwh I reduced App.jsx to this:

```jsx
import * as React from 'react';

export function App() {
  return (
    <div>
    This is App.jsx
    </div>
  );
}
```

the result (via `npm build` removing minification):

```js
/* 10 */
/***/ (function(module, exports, __webpack_require__) {

"use strict";


var _interopRequireWildcard = __webpack_require__(1);

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.App = App;

var React = _interopRequireWildcard(__webpack_require__(0));

function App() {
  return /*#__PURE__*/React.createElement("div", null, "This is App.jsx");
}

_c = App;

var _c;

$RefreshReg$(_c, "App");

/***/ })
```

**Comment by michaelgmiller1** [comment]
I reduced `App.jsx` to a minimum to reproduce:

```jsx
import * as React from 'react';
import ClassDefault from './ClassDefault';

export function App() {
  return (
    <div>
      This is App.jsx!!!!!!
    </div>
  );
}
```

The result is:

```jsx
/* 11 */
/***/ (function(module, exports, __webpack_require__) {

"use strict";


var _interopRequireDefault = __webpack_require__(4);

var _interopRequireWildcard = __webpack_require__(1);

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.App = App;

var React = _interopRequireWildcard(__webpack_require__(0));

var _ClassDefault = _interopRequireDefault(__webpack_require__(12));

function App() {
  return /*#__PURE__*/React.createElement("div", null, "This is App.jsx!!!!!!");
}

_c = App;

var _c;

$RefreshReg$(_c, "App");

/***/ }),
```

**Comment by michaelgmiller1** [comment]
So it looks like it does:

```
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.App = App;
``

on the app component. Should that interfere with hot reloading?

**Comment by pmmmwh** [comment]
I did a bit of digging and it seems that this only happens when `@babel/plugin-transform-modules-commonjs` and `@babel/plugin-transform-runtime` is both used.

The culprit of this issue is that somehow the referential equality of the object which we use to calculate whether the change is acceptable is broken - I'll have to look further (and maybe even ping Dan) to get this properly fixed.

**Comment by pmmmwh** [comment]
@michaelgmiller1 I think I've found the culprit.

The issue in the repro is that there are two instances of `react-refresh/runtime` being used, and thus object referential equality will be broken across the two instances. I haven't figured out why the combination of the two plugins will cause this though - it might have something to do with helper injection. (Also unsure why the issue didn't surface earlier ...)

After making sure that only one copy is loaded, I seem to have successful results.
[See CodeSandbox example here](https://codesandbox.io/s/babel-transform-react-refresh-lx9j8)

**Comment by michaelgmiller1** [comment]
@pmmmwh thanks for diving in! just one comment on your codebox, it uses `export default App` instead of `export function App`. Exporting via object was needed to reproduce the issue for me. If I change your code sample to export via object, then it breaks again.

**Comment by pmmmwh** [comment]
> @pmmmwh thanks for diving in! just one comment on your codebox, it uses `export default App` instead of `export function App`. Exporting via object was needed to reproduce the issue for me. If I change your code sample to export via object, then it breaks again.

This ... is ... so ... strange!

I've tested with a bare bones example on my local machine with `export function App` and it works 😱

**Comment by pmmmwh** [comment]
Uhh ... okay. I have something working. I will add a regression test and some notes in the code base.

**Comment by pmmmwh** [comment]
RE-FIXED in `0.4.0-beta.8`.
